File: windows/ScreenTimeGuardian/LLMRanking.cs
using System.Globalization;
using System.Net.Http.Json;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace ScreenTimeGuardian;

internal sealed class LLMRankingCache
{
    [JsonPropertyName("week_id")]
    public string WeekId { get; set; } = "";

    [JsonPropertyName("period_start")]
    public string? PeriodStart { get; set; }

    [JsonPropertyName("period_end")]
    public string? PeriodEnd { get; set; }

    [JsonPropertyName("source")]
    public string Source { get; set; } = "openrouter.ai";

    [JsonPropertyName("fetched_at_utc")]
    public DateTimeOffset FetchedAtUtc { get; set; }

    [JsonPropertyName("rows")]
    public List<LLMRankingRow> Rows { get; set; } = [];
}

internal sealed class LLMRankingRow
{
    [JsonPropertyName("rank")]
    public int Rank { get; set; }

    [JsonPropertyName("llm_name")]
    public string LLMName { get; set; } = "";

    [JsonPropertyName("prompt_tokens")]
    public double PromptTokens { get; set; }

    [JsonPropertyName("output_tokens")]
    public double OutputTokens { get; set; }

    [JsonPropertyName("weighted_average_input_price")]
    public double WeightedAverageInputPrice { get; set; }

    [JsonPropertyName("weighted_average_output_price")]
    public double WeightedAverageOutputPrice { get; set; }

    [JsonPropertyName("weekly_revenue")]
    public double WeeklyRevenue { get; set; }
}

internal sealed class OpenRouterClient
{
    private static readonly HttpClient Client = new()
    {
        BaseAddress = new Uri("https://openrouter.ai"),
        Timeout = TimeSpan.FromSeconds(30)
    };

    public async Task<LLMRankingCache> FetchWeeklyPromptTokenTop20Async(CancellationToken token = default)
    {
        var entries = await FetchRankingEntriesAsync(token);
        var topEntries = entries
            .Where(row => !string.IsNullOrWhiteSpace(row.Permaslug) && row.PromptTokens > 0)
            .OrderByDescending(row => row.PromptTokens)
            .Take(20)
            .ToList();

        var rows = new List<LLMRankingRow>();
        var rank = 1;
        foreach (var entry in topEntries)
        {
            var pricing = await FetchEffectivePricingOrDefaultAsync(entry.Permaslug, entry.Variant, token);
            var revenue =
                entry.PromptTokens / 1_000_000 * pricing.WeightedInputPricePerMillion
                + entry.OutputTokens / 1_000_000 * pricing.WeightedOutputPricePerMillion;
            rows.Add(new LLMRankingRow
            {
                Rank = rank++,
                LLMName = entry.Permaslug,
                PromptTokens = entry.PromptTokens,
                OutputTokens = entry.OutputTokens,
                WeightedAverageInputPrice = pricing.WeightedInputPricePerMillion,
                WeightedAverageOutputPrice = pricing.WeightedOutputPricePerMillion,
                WeeklyRevenue = revenue
            });
        }

        var weekStart = DateTools.CurrentWeekStart();
        return new LLMRankingCache
        {
            WeekId = DateTools.WeekId(),
            PeriodStart = DateTools.DateString(weekStart),
            PeriodEnd = DateTools.DateString(weekStart.AddDays(6)),
            Source = "openrouter.ai/api/frontend/v1",
            FetchedAtUtc = DateTimeOffset.UtcNow,
            Rows = rows
        };
    }

    private static async Task<List<OpenRouterRankingEntry>> FetchRankingEntriesAsync(CancellationToken token)
    {
        using var response = await Client.GetAsync("/api/frontend/v1/rankings/models?view=week", token);
        response.EnsureSuccessStatusCode();
        using var document = await JsonDocument.ParseAsync(await response.Content.ReadAsStreamAsync(token), cancellationToken: token);
        if (!document.RootElement.TryGetProperty("data", out var data) || data.ValueKind != JsonValueKind.Array)
        {
            throw new InvalidOperationException("OpenRouter rankings 返回格式缺少 data 数组");
        }

        var entries = new List<OpenRouterRankingEntry>();
        foreach (var row in data.EnumerateArray())
        {
            var permaslug = StringValue(row, "model_permaslug");
            if (string.IsNullOrWhiteSpace(permaslug))
            {
                continue;
            }

            entries.Add(new OpenRouterRankingEntry(
                permaslug,
                string.IsNullOrWhiteSpace(StringValue(row, "variant")) ? "standard" : StringValue(row, "variant"),
                NumberValue(row, "total_prompt_tokens"),
                NumberValue(row, "total_completion_tokens")));
        }

        if (entries.Count == 0)
        {
            throw new InvalidOperationException("OpenRouter rankings 没有返回可用数据");
        }

        return entries;
    }

    private static async Task<OpenRouterEffectivePricing> FetchEffectivePricingOrDefaultAsync(string permaslug, string variant, CancellationToken token)
    {
        try
        {
            var url = $"/api/frontend/v1/stats/effective-pricing?permaslug={Uri.EscapeDataString(permaslug)}&variant={Uri.EscapeDataString(string.IsNullOrWhiteSpace(variant) ? "standard" : variant)}";
            using var response = await Client.GetAsync(url, token);
            response.EnsureSuccessStatusCode();
            var payload = await response.Content.ReadFromJsonAsync<JsonElement>(cancellationToken: token);
            if (!payload.TryGetProperty("data", out var data))
            {
                return new OpenRouterEffectivePricing(0, 0);
            }

            return new OpenRouterEffectivePricing(
                NumberValue(data, "weightedInputPrice"),
                NumberValue(data, "weightedOutputPrice"));
        }
        catch
        {
            return new OpenRouterEffectivePricing(0, 0);
        }
    }

    private static string StringValue(JsonElement row, string property)
    {
        return row.TryGetProperty(property, out var value) && value.ValueKind == JsonValueKind.String
            ? value.GetString() ?? ""
            : "";
    }

    private static double NumberValue(JsonElement row, string property)
    {
        if (!row.TryGetProperty(property, out var value))
        {
            return 0;
        }

        return value.ValueKind switch
        {
            JsonValueKind.Number => value.GetDouble(),
            JsonValueKind.String when double.TryParse(value.GetString(), NumberStyles.Any, CultureInfo.InvariantCulture, out var result) => result,
            _ => 0
        };
    }
}

internal sealed record OpenRouterRankingEntry(string Permaslug, string Variant, double PromptTokens, double OutputTokens);

internal sealed record OpenRouterEffectivePricing(double WeightedInputPricePerMillion, double WeightedOutputPricePerMillion);

File: android/app/src/main/java/com/timbertrail/screentimeguardian/LLMRanking.kt
package com.timbertrail.screentimeguardian

import org.json.JSONArray
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder
import java.time.Instant
import java.time.LocalDate
import java.util.Locale

data class LLMRankingCache(
    val weekId: String,
    val periodStart: String?,
    val periodEnd: String?,
    val source: String,
    val fetchedAtUtc: Instant,
    val rows: List<LLMRankingRow>
) {
    fun toJson(): JSONObject {
        val array = JSONArray()
        rows.forEach { array.put(it.toJson()) }
        return JSONObject()
            .put("week_id", weekId)
            .put("period_start", periodStart)
            .put("period_end", periodEnd)
            .put("source", source)
            .put("fetched_at_utc", fetchedAtUtc.toString())
            .put("rows", array)
    }

    companion object {
        fun fromJson(json: JSONObject): LLMRankingCache {
            val rowsArray = json.optJSONArray("rows") ?: JSONArray()
            return LLMRankingCache(
                weekId = json.optString("week_id", DateTools.weekId()),
                periodStart = json.optString("period_start").takeIf { it.isNotBlank() && it != "null" },
                periodEnd = json.optString("period_end").takeIf { it.isNotBlank() && it != "null" },
                source = json.optString("source", "openrouter.ai"),
                fetchedAtUtc = DateTools.parseInstant(json.optString("fetched_at_utc", Instant.now().toString())),
                rows = (0 until rowsArray.length()).map { LLMRankingRow.fromJson(rowsArray.getJSONObject(it)) }
            )
        }
    }
}

data class LLMRankingRow(
    val rank: Int,
    val llmName: String,
    val promptTokens: Double,
    val outputTokens: Double,
    val weightedAverageInputPrice: Double,
    val weightedAverageOutputPrice: Double,
    val weeklyRevenue: Double
) {
    fun toJson(): JSONObject = JSONObject()
        .put("rank", rank)
        .put("llm_name", llmName)
        .put("prompt_tokens", promptTokens)
        .put("output_tokens", outputTokens)
        .put("weighted_average_input_price", weightedAverageInputPrice)
        .put("weighted_average_output_price", weightedAverageOutputPrice)
        .put("weekly_revenue", weeklyRevenue)

    companion object {
        fun fromJson(json: JSONObject): LLMRankingRow = LLMRankingRow(
            rank = json.optInt("rank", 0),
            llmName = json.optString("llm_name"),
            promptTokens = json.optDouble("prompt_tokens", 0.0),
            outputTokens = json.optDouble("output_tokens", 0.0),
            weightedAverageInputPrice = json.optDouble("weighted_average_input_price", 0.0),
            weightedAverageOutputPrice = json.optDouble("weighted_average_output_price", 0.0),
            weeklyRevenue = json.optDouble("weekly_revenue", 0.0)
        )
    }
}

private data class OpenRouterRankingEntry(
    val permaslug: String,
    val variant: String,
    val promptTokens: Double,
    val outputTokens: Double
)

private data class OpenRouterEffectivePricing(
    val weightedInputPricePerMillion: Double,
    val weightedOutputPricePerMillion: Double
)

class OpenRouterClient {
    fun fetchWeeklyPromptTokenTop20(): LLMRankingCache {
        val entries = fetchRankingEntries()
            .filter { it.permaslug.isNotBlank() && it.promptTokens > 0 }
            .sortedByDescending { it.promptTokens }
            .take(20)

        val rows = entries.mapIndexed { index, entry ->
            val pricing = runCatching { fetchEffectivePricing(entry.permaslug, entry.variant) }
                .getOrDefault(OpenRouterEffectivePricing(0.0, 0.0))
            val revenue =
                entry.promptTokens / 1_000_000.0 * pricing.weightedInputPricePerMillion +
                    entry.outputTokens / 1_000_000.0 * pricing.weightedOutputPricePerMillion
            LLMRankingRow(
                rank = index + 1,
                llmName = entry.permaslug,
                promptTokens = entry.promptTokens,
                outputTokens = entry.outputTokens,
                weightedAverageInputPrice = pricing.weightedInputPricePerMillion,
                weightedAverageOutputPrice = pricing.weightedOutputPricePerMillion,
                weeklyRevenue = revenue
            )
        }

        val weekStart = DateTools.currentWeekStartDate()
        return LLMRankingCache(
            weekId = DateTools.weekId(weekStart),
            periodStart = DateTools.dateString(weekStart),
            periodEnd = DateTools.dateString(weekStart.plusDays(6)),
            source = "openrouter.ai/api/frontend/v1",
            fetchedAtUtc = Instant.now(),
            rows = rows
        )
    }

    private fun fetchRankingEntries(): List<OpenRouterRankingEntry> {
        val json = fetchJson("https://openrouter.ai/api/frontend/v1/rankings/models?view=week")
        val data = json.optJSONArray("data") ?: error("OpenRouter rankings 返回格式缺少 data 数组")
        return (0 until data.length()).mapNotNull { index ->
            val row = data.optJSONObject(index) ?: return@mapNotNull null
            val permaslug = row.optString("model_permaslug")
            if (permaslug.isBlank()) return@mapNotNull null
            OpenRouterRankingEntry(
                permaslug = permaslug,
                variant = row.optString("variant").ifBlank { "standard" },
                promptTokens = row.optFlexibleDouble("total_prompt_tokens"),
                outputTokens = row.optFlexibleDouble("total_completion_tokens")
            )
        }
    }

    private fun fetchEffectivePricing(permaslug: String, variant: String): OpenRouterEffectivePricing {
        val url =
            "https://openrouter.ai/api/frontend/v1/stats/effective-pricing?permaslug=${permaslug.urlEncoded()}&variant=${variant.ifBlank { "standard" }.urlEncoded()}"
        val data = fetchJson(url).optJSONObject("data") ?: return OpenRouterEffectivePricing(0.0, 0.0)
        return OpenRouterEffectivePricing(
            weightedInputPricePerMillion = data.optFlexibleDouble("weightedInputPrice"),
            weightedOutputPricePerMillion = data.optFlexibleDouble("weightedOutputPrice")
        )
    }

    private fun fetchJson(url: String): JSONObject {
        val connection = URL(url).openConnection() as HttpURLConnection
        connection.connectTimeout = 30_000
        connection.readTimeout = 30_000
        connection.requestMethod = "GET"
        connection.setRequestProperty("Accept", "application/json")
        connection.inputStream.bufferedReader().use { reader ->
            return JSONObject(reader.readText())
        }
    }
}

private fun JSONObject.optFlexibleDouble(name: String): Double {
    val value = opt(name) ?: return 0.0
    return when (value) {
        is Number -> value.toDouble()
        is String -> value.toDoubleOrNull() ?: 0.0
        else -> 0.0
    }
}

private fun String.urlEncoded(): String =
    URLEncoder.encode(this, Charsets.UTF_8.name())

fun formatNumber(value: Double): String = String.format(Locale.US, "%,.0f", value)

fun formatPrice(value: Double): String = String.format(Locale.US, "$%,.2f", value)

fun formatWholeCurrency(value: Double): String = String.format(Locale.US, "$%,.0f", value)
